src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+        for pulse in sequence.pulses:
+            if pulse.channel not in self._ports:
+                continue
+
+            dac = self.ports(pulse.channel).dac
+            start = int(pulse.start * 1e-9 * dac_sampling_rate)
+            i_env = pulse.envelope_waveform_i(dac_sr_ghz).data
+            q_env = pulse.envelope_waveform_q(dac_sr_ghz).data
+
+            # Flux pulses
+            # TODO: Add envelope support for flux pulses
+            if pulse.type == PulseType.FLUX:
+                wfm = i_env
+                end = start + len(wfm)
+
+            # Qubit drive microwave signals
+            elif pulse.type == PulseType.DRIVE:
+                end = start + len(i_env)
+                t = np.arange(start, end) / dac_sampling_rate
+                cosalpha = np.cos(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                sinalpha = np.sin(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                wfm = i_env * sinalpha + q_env * cosalpha
+
+            elif pulse.type == PulseType.READOUT:
+                # For readout pulses, we move the corresponding DAC/ADC pair to the start of the pulse to save memory
+                # This locks the phase of the readout in the demodulation
+                adc = self.ports(pulse.channel).adc
+                start = 0
+
+                end = start + len(i_env)
+                t = np.arange(start, end) / dac_sampling_rate
+                cosalpha = np.cos(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                sinalpha = np.sin(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                wfm = i_env * sinalpha + q_env * cosalpha
+
+                # First we convert the pulse starting time to number of ADC samples
+                # Then, we convert this number to the number of ADC clock cycles (8 samples per clock cycle)
+                # Next, we raise it to the next nearest integer to prevent an overlap between drive and readout pulses
+                # Finally, we ensure that the number is even for the DAC delay conversion
+                delay_start_adc = int(
+                    int(
+                        np.ceil(
+                            self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9 / 8
+                        )
+                        / 2
+                    )
+                    * 2
+                )
 
-    def stop(self):
-        self._send(b"stop")
+                # For the DAC, currently the sampling rate is 3x higher than the ADC
+                # The number of clock cycles is 16 samples per clock cycle
+                # Hence, we multiply the adc delay clock cycles by 1.5x to align the DAC/ADC pair
+                delay_start_dac = int(delay_start_adc * 1.5)
```

Comment:
Despite the comments, it's still a bit confusing. I believe it would be beneficial to spell out units.

My understanding, it's that you what to find the number of clock cycles to wait before start both ADC and DAC (as written in the last comment line).

So, I'd write something like:
```py
delay_samples_adc = self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9  # ADC samples to skip
_base = int(np.ceil(delay_samples_adc / 8) / 2) # cycles "base"
delay_start_adc = _base * 2  # ADC clock cycles
delay_start_dac = _base * 3  # DAC clock cycles, different by ADC
```

The thing that is not fully clear is what "base" does mean. The most straightforward would be to use instead directly samples per clock cycle `dacscc = 16` and `adcscc = 8` and `lcm = lcm(dacscc, adcscc) = 16` and the sampling rates
```py
_base = np.ceil(delay_samples_adc / lcm)
delay_start_adc = _base * lcm / adcscc  # ADC clock cycles
delay_start_dac = _base * lcm / dacscc * dacsr / adcsr  # DAC clock cycles, different by ADC
```
since `dacsr / adcsr` is integer (otherwise you should look for the `lcm(dacscc * adcsr, adcscc * dacsr)` directly). Though the full thing looks a bit more verbose (and it's for sure less efficient)

However, moving part of the factor between the `ceil` and the outside `int` is not fully equivalent...